Add support for linting and scoring dbt seeds

StrahilPeykov opened the pull request:
## Overview
Add support for linting and scoring seed resources in dbt-score, following issue [#105](https://github.com/PicnicSupermarket/dbt-score/issues/105).

## Problem
Previously, dbt-score only supported linting models, sources, and snapshots. Seeds were not evaluated, creating an inconsistency in the quality assessment of a dbt project's metadata. Since seeds often contain important reference data, ensuring they have proper documentation and ownership is valuable.

## Implementation
- Added `Seed` class to represent dbt seeds
- Updated `ManifestLoader` to load seeds from the manifest
- Added seed-specific linting rules (description, columns, tests, ownership)
- Updated `Evaluation` class to include seeds in evaluation chain
- Modified formatters to handle and display seed results
- Added comprehensive tests for seed support

## New Rules
- `seed_has_description` - Ensures seeds have descriptive documentation
- `seed_columns_have_description` - Verifies seed columns are documented
- `seed_has_tests` - Checks that seeds have appropriate tests
- `seed_has_owner` - Ensures seeds have defined ownership

## Testing
Full test coverage has been added for seed support.
- Fixtures for seeds in test suite
- Tests for seed-specific rules
- Updates to existing tests to accommodate seeds

jochemvandooren commented:
Thanks for opening a PR, I will have a look soon. There's some linting errors by the way: https://github.com/PicnicSupermarket/dbt-score/actions/runs/14473008011/job/40645749304?pr=110

jochemvandooren commented on CHANGELOG.md:
```diff
@@ -8,6 +8,8 @@ and this project adheres to
 
 ## [Unreleased]
 
+- Add support for linting and scoring dbt seeds (#104)
```
```suggestion
- Add support for linting and scoring dbt seeds (#110)
```

jochemvandooren commented on docs/seed.md:
Although I like this idea, we have to keep things consistent: `model`, `source` and `snapshot` do not have such a file, so I'd rather leave it out for now, we can add this in a follow-up PR?

jochemvandooren commented on src/dbt_score/rules/generic.py:
```diff
@@ -134,3 +134,37 @@ def has_no_unused_is_incremental(model: Model) -> RuleViolation | None:
+    if invalid_column_names:
+        max_length = 60
+        message = f"Columns lack a description: {', '.join(invalid_column_names)}."
+        if len(message) > max_length:
```
Isn't this redundant, as you can also do `f"{message[:60]}…"` if the length is lower than 60? It will just show the full string I think

jochemvandooren commented on src/dbt_score/rules/generic.py:
```diff
@@ -134,3 +134,37 @@ def has_no_unused_is_incremental(model: Model) -> RuleViolation | None:
+        return RuleViolation(message=message)
+
+
+@rule
+def seed_has_tests(seed: Seed) -> RuleViolation | None:
+    """A seed should have at least one test."""
+    if not seed.tests and not any(column.tests for column in seed.columns):
+        return RuleViolation(message="Seed has no tests.")
```
I am not sure if I'd like to enforce this rule on every user. We can also make it a low severity?

jochemvandooren reviewed (commented):
Thanks for your contribution! :raised_hands: Overall, the feature looks very good, I will have a closer look at the tests tomorrow. Left some small comments already

jochemvandooren commented:
Also, there's still some linting errors related to `mypy`! You can run `pre-commit run --all-files` locally to get those errors

StrahilPeykov commented:
Thanks a lot for the feedback @jochemvandooren! I've addressed all your comments - updated the PR number in the CHANGELOG, removed the seed.md documentation file for consistency, simplified the max_length check in the column description rule, changed the severity of `seed_has_tests` to LOW, and fixed all the mypy and linting issues with pre-commit.

jochemvandooren commented on tests/test_cli.py:
```diff
@@ -19,12 +20,15 @@ def test_invalid_options():
 def test_lint_existing_manifest(manifest_path):
     """Test lint with an existing manifest."""
-    with patch("dbt_score.cli.Config._load_toml_file"):
+    with patch("dbt_score.cli.lint_dbt_project") as mock_lint:
```
Why do we do this here? Now we are patching `lint_dbt_project`, meaning we will not actually lint the manifest, which was the goal of the test.

jochemvandooren commented on tests/test_cli.py:
```diff
@@ -93,26 +100,38 @@ def test_lint_other_exception(manifest_path, caplog):
+    mock_eval.project_score = Score(5.0, "🥉")  # Score below 10.0
+    mock_eval.scores.values.return_value = []
+
+    with patch("dbt_score.cli.lint_dbt_project") as mock_lint:
```
I think here it makes more sense! Now we actually test *only* the `fail_project_under` behavior :+1:

jochemvandooren commented on tests/test_cli.py:
```diff
@@ -93,26 +100,38 @@ def test_lint_other_exception(manifest_path, caplog):
+
+    with patch("dbt_score.cli.lint_dbt_project") as mock_lint:
+        mock_lint.return_value = mock_eval
+        # Also patch the HumanReadableFormatter to control the output
```
Stray comment?

jochemvandooren commented on tests/test_cli.py:
```diff
@@ -93,26 +100,38 @@ def test_lint_other_exception(manifest_path, caplog):
         assert result.exit_code == 1
 
 
 def test_fail_any_model_under(manifest_path):
```
```suggestion
def test_fail_any_item_under(manifest_path):
```
Consistency :nerd_face:

jochemvandooren commented on tests/test_seed_rules.py:
This is great :ok_hand: We should do the same for other dbt entities!

jochemvandooren reviewed (commented):
Just left some final comments on the tests, great work and thanks for improving the tests! :raised_hands:

StrahilPeykov commented on src/dbt_score/rules/generic.py:
```diff
@@ -134,3 +136,37 @@ def has_no_unused_is_incremental(model: Model) -> RuleViolation | None:
+    ]
+    if invalid_column_names:
+        message = f"Columns lack a description: {', '.join(invalid_column_names)}."
+        if len(message) > MAX_DESCRIPTION_MESSAGE_LENGTH:
+            message = f"{message[:MAX_DESCRIPTION_MESSAGE_LENGTH]}…"
```
I addressed all the other issues in the latest commits. Regarding the constant, I prefer to keep it that way rather than hardcoding it, because if we ever wanna change the length limit, it's easier to just update it in one place, but if you prefer a hardcoded approach here that's fine with me

matthieucan replied:
I believe the code is more readable if all the needed context to understand a function is close to or within that function. Adding an indirection makes it harder to read in this case, and this function will be read much more often than the constant will be changed :) So we should simplify it!

jochemvandooren replied:
Let's finalize this discussion as well :grin:

StrahilPeykov replied:
I still think that even if hardcoding might be a bit more readable, using a constant is better for maintainability and also satisfies the linting rules (PLR2004 rule for no magic numbers). But I'll leave it up to you to decide:)

StrahilPeykov replied:
done:)

jochemvandooren reviewed (approved):
Nice, thanks a lot! :raised_hands: I'll leave the last discussion up to you and @matthieucan

jochemvandooren commented on CHANGELOG.md:
```diff
@@ -8,6 +8,8 @@ and this project adheres to
 ## [Unreleased]
 
+- Add support for linting and scoring dbt seeds (#110)
+
```
```suggestion
```

jochemvandooren commented on src/dbt_score/models.py:
```diff
@@ -594,3 +712,56 @@ def _filter_evaluables(self, select: Iterable[str]) -> None:
+                return seed
+        return None
+
+    def get_first_model(self) -> Model | None:
+        """Get the first model in the collection, if any."""
+        return next(iter(self.models.values())) if self.models else None
+
+    def get_first_source(self) -> Source | None:
+        """Get the first source in the collection, if any."""
+        return next(iter(self.sources.values())) if self.sources else None
+
+    def get_first_snapshot(self) -> Snapshot | None:
+        """Get the first snapshot in the collection, if any."""
+        return next(iter(self.snapshots.values())) if self.snapshots else None
+
+    def get_first_seed(self) -> Seed | None:
+        """Get the first seed in the collection, if any."""
+        return next(iter(self.seeds.values())) if self.seeds else None
```
I don't think these methods serve any purpose, other than being used in the tests. So I suggest not creating these

jochemvandooren replied:
Actually the case for all these helper functions

jochemvandooren commented on src/dbt_score/models.py:
```diff
@@ -550,6 +641,44 @@ def _load_snapshots(self) -> None:
+        elif parent_id in self.seeds:
+            node.parents.append(self.seeds[parent_id])
+
+    def _populate_parents(self) -> None:
```
Why was this method changed? I try to keep the changes related to seeds and not change to much related to other functionalities to keep things small and related to a single feature! Also this code was reviewed, approved and merged so I see no reason to change it unless you have very good reasons to, does that make sense?

StrahilPeykov replied:
Yeah you're right, I added them for convenience for tests, but I have now removed them

jochemvandooren reviewed (commented):
Great to see the rebasing worked out :ok_hand: I suggest try keeping the changes related to the seed feature only! Left a couple of comments about that

jochemvandooren commented on tests/test_evaluation.py:
```diff
@@ -85,8 +95,11 @@ def test_evaluation_critical(
 
     evaluation.evaluate()
 
-    model2 = manifest_loader.models["model.package.model2"]
```
i don't see what's wrong with fetching the model from the manifest by it's ID? I think this is a neater way of doing it than having to search it by name?

jochemvandooren replied:
Same for the other ocurrences, i think fetching the model by it's key should be the best way to do it!

jochemvandooren commented on tests/test_models.py:
```diff
@@ -79,3 +82,74 @@ def test_manifest_no_model(mock_dbt_ls, mock_read_text, raw_manifest, caplog):
+
+
+@patch("dbt_score.models.Path.read_text")
+def test_parent_references(mock_read_text, raw_manifest):
```
I think a test is already in place for this: https://github.com/ross-whatnot/dbt-score/blob/edf0563d0f93aecd41a639ccdb628eff5ae8aded/tests/test_models.py#L38-L49

jochemvandooren reviewed (commented):
Some final comment! :raised_hands:

matthieucan commented on src/dbt_score/rules/generic.py:
```diff
@@ -1,8 +1,10 @@
+from dbt_score import Model, RuleViolation, Seed, Severity, Snapshot, rule
 from dbt_score.rules.filters import is_table
 
+MAX_DESCRIPTION_MESSAGE_LENGTH = 60
```
This is now unused, right?

StrahilPeykov replied:
Yes, it is removed now
